src/Http/Middleware/EnsureDeviceHasNotBeenLoggedOut.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Laravel\Sanctum\Http\Middleware;
+
+use Closure;
+use Illuminate\Auth\AuthenticationException;
+use Illuminate\Auth\SessionGuard;
+use Illuminate\Contracts\Auth\Factory as AuthFactory;
+use Illuminate\Http\Request;
+use Illuminate\Support\Arr;
+use Symfony\Component\HttpFoundation\Response;
+
+class EnsureDeviceHasNotBeenLoggedOut
+{
+    /**
+     * The authentication factory implementation.
+     *
+     * @var \Illuminate\Contracts\Auth\Factory
+     */
+    protected $auth;
+
+    /**
+     * Create a new middleware instance.
+     *
+     * @param  \Illuminate\Contracts\Auth\Factory  $auth
+     * @return void
+     */
+    public function __construct(AuthFactory $auth)
+    {
+        $this->auth = $auth;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Closure(\Illuminate\Http\Request): (\Symfony\Component\HttpFoundation\Response)  $next
+     *
+     * @throws \Illuminate\Auth\AuthenticationException
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        if ($request->hasSession()) {
+            if ($request->session()->has($key = 'password_hash_'.$this->auth->getDefaultDriver())) {
```

Comment:
From my test `$this->auth->getDefaultDriver()` always returns `sanctum`, using this logic we can loop `sanctum.guard` config and only logged out if the `$guard` in an instance of `SessionGuard`.

Reply:
From the [test repo](https://github.com/patrickomeara/sanctum-logout-issue) I have setup, `$this->auth->getDefaultDriver()` returns `web` on a sanctum request, but will use `sanctum` after the request has processed, as the request changes the default guard via `shouldUse()`.

Once we have added the `tap()` to store the password hash after the request as mentioned in my other comment, we now have two different guards being used. `web` and `sanctum`.

Ray screenshot:
![image](https://github.com/laravel/sanctum/assets/571773/d27007f2-43e5-4ae3-b872-a07be6887bb8)

We can get around this by caching the password hash key.

```php

    protected function passwordHashKey()
    {
        return $this->passwordHashKey ??= 'password_hash_'.$this->auth->getDefaultDriver();
    }
```

Reply:
Changed the middleware to only check for auth guards from `sanctum.guard` on only logged out or store password hash on `SessionGuard`.